src/wallet/wallet.cpp

```diff
@@ -2194,19 +2204,23 @@ void UpdateWitnessHeights(NoteDataMap& noteDataMap, int indexHeight, int64_t nWi
     }
 }
 
-void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
-                                     const CBlock* pblockIn,
-                                     SproutMerkleTree& sproutTree,
-                                     SaplingMerkleTree& saplingTree)
+void CWallet::IncrementNoteWitnesses(
+        const Consensus::Params& consensus,
+        const CBlockIndex* pindex,
+        const CBlock* pblockIn,
+        SproutMerkleTree& sproutTree,
+        SaplingMerkleTree& saplingTree,
+        bool performOrchardWalletUpdates)
 {
     LOCK(cs_wallet);
     for (std::pair<const uint256, CWalletTx>& wtxItem : mapWallet) {
        ::CopyPreviousWitnesses(wtxItem.second.mapSproutNoteData, pindex->nHeight, nWitnessCacheSize);
        ::CopyPreviousWitnesses(wtxItem.second.mapSaplingNoteData, pindex->nHeight, nWitnessCacheSize);
     }
 
-    // ORCHARD: Add a checkpoint before we add information from the new block
-    orchardWallet.CheckpointNoteCommitmentTree();
+    if (consensus.NetworkUpgradeActive(pindex->nHeight, Consensus::UPGRADE_NU5)) {
+        assert(!performOrchardWalletUpdates || orchardWallet.CheckpointNoteCommitmentTree(pindex->nHeight));
+    }
```

Comment:
```suggestion
    if (performOrchardWalletUpdates && consensus.NetworkUpgradeActive(pindex->nHeight, Consensus::UPGRADE_NU5)) {
        assert(orchardWallet.CheckpointNoteCommitmentTree(pindex->nHeight));
    }
```
This is equivalent but I think slightly clearer, since it doesn't rely on shortcutting to control whether a side effect is performed.